packages/internal/src/utils.js

```diff
@@ -348,3 +349,104 @@ export const allValues = async obj => {
   // @ts-expect-error cast
   return harden(fromEntries(zip(keys(obj), resolved)));
 };
+
+/**
+ * A tee implementation where all readers are synchronized with each other.
+ * They all consume the source stream in lockstep, and any one returning or
+ * throwing early will affect the others.
+ *
+ * @template [T=unknown]
+ * @param {AsyncIterator<T, void, void>} sourceStream
+ * @param {number} readerCount
+ */
+export const synchronizedTee = (sourceStream, readerCount) => {
+  /** @type {IteratorReturnResult<void> | undefined} */
+  let doneResult;
+
+  /** @typedef {IteratorResult<(value: PromiseLike<IteratorResult<T>>) => void>} QueuePayload */
+  /** @type {import('@endo/stream').AsyncQueue<QueuePayload>[]} */
+  const queues = [];
+
+  /** @returns {Promise<void>} */
+  const pullNext = async () => {
+    const requests = await Promise.allSettled(queues.map(queue => queue.get()));
+    const rejections = [];
+    /** @type {Array<(value: PromiseLike<IteratorResult<T>>) => void>} */
+    const resolvers = [];
+    let done = false;
+    for (const settledResult of requests) {
+      if (settledResult.status === 'rejected') {
+        rejections.push(settledResult.reason);
+      } else {
+        done ||= !!settledResult.value.done;
+        resolvers.push(settledResult.value.value);
+      }
+    }
+    /** @type {Promise<IteratorResult<T>>} */
+    let result;
+    if (doneResult) {
+      result = Promise.resolve(doneResult);
+    } else if (rejections.length) {
+      const error = assert.error(assert.details`Teed stream threw`);
+      assert.note(error, assert.details`Teed rejections: ${rejections}`);
+      result =
+        sourceStream.throw?.(error) ||
+        Promise.resolve(sourceStream.return?.()).then(() =>
+          Promise.reject(error),
+        );
+    } else if (done) {
+      result =
+        sourceStream.return?.() ||
```

Comment:
Suggest `??` over `||` here.

Reply:
Once you have an array of promises that you know have settled, some of which may be rejected, it might be sufficient to lean on `Promise.all` to consolidate them into a rejected promise with an `AggregateError` instead of rolling that by hand.

But, it would suffice here to leave a comment with the reason you did not do that, which appears to be to connect with SES causal tracing.

Reply:
Tracking taming Promise.all, worth referencing too https://github.com/endojs/endo/issues/1568

Reply:
Promise.all short circuits on rejection, which mean aggregation is not possible. I have in fact a `PromiseAllOrRejections` helper that does not short circuit, but alas it's not standard (yet?)